lib/compute/agent-node-config.ts

```diff
@@ -6,39 +6,108 @@
-        cpuType: AmazonLinuxCpuType.X86_64,
-      }).getImage(stack).imageId.toString(),
-    };
-    this.AL2_ARM64 = {
-      ec2CloudName: 'AL2-ARM64',
-      instanceType: InstanceType.of(InstanceClass.C6G, InstanceSize.XLARGE4).toString(),
-      workerLabelString: 'AL2-ARM64',
-      numberOfExecutors: '2',
-      remoteUser: 'ec2-user',
-      amiId: MachineImage.latestAmazonLinux({
-        generation: AmazonLinuxGeneration.AMAZON_LINUX_2,
-        cpuType: AmazonLinuxCpuType.ARM_64,
-      }).getImage(stack).imageId.toString(),
+    this.STACKREGION = stack.region;
+    const key = new KeyPair(stack, 'AgentNode-KeyPair', {
+      name: 'AgentNodeKeyPair',
+      description: 'KeyPair used by Jenkins Main Node to SSH into Agent Nodes',
+    });
+    Tags.of(key)
+      .add('jenkins:credentials:type', 'sshUserPrivateKey');
+    const AgentNodeRole = new Role(stack, 'JenkinsAgentNodeRole', {
+      assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
+    });
+    const AgentNodeInstanceProfile = new CfnInstanceProfile(stack, 'JenkinsAgentNodeInstanceProfile', { roles: [AgentNodeRole.roleName] });
+    this.AgentNodeInstanceProfileArn = AgentNodeInstanceProfile.attrArn.toString();
+    this.SSHEC2KeySecretId = Fn.join('/', ['ec2-ssh-key', key.keyPairName.toString(), 'private']);
+  }
+
+  public addAgentConfigToJenkinsYaml(templates: AgentNodeProps[], props: AgentNodeNetworkProps): any {
+    const jenkinsYaml: any = load(readFileSync(JenkinsMainNode.BASE_JENKINS_YAML_PATH, 'utf-8'));
+    const configTemplates: any = [];
+
+    templates.forEach((element) => {
+      configTemplates.push(this.getTemplate(element, props));
+    });
+
+    const agentNodeYamlConfig = [{
+      amazonEC2: {
+        cloudName: 'Amazon_ec2_cloud',
+        region: this.STACKREGION,
+        sshKeysCredentialsId: this.SSHEC2KeySecretId,
+        templates: configTemplates,
+        useInstanceProfileForCredentials: true,
+      },
+    }];
+    jenkinsYaml.jenkins.clouds = agentNodeYamlConfig;
+    return jenkinsYaml;
+  }
+
+  private getTemplate(config: AgentNodeProps, props: AgentNodeNetworkProps): { [x: string]: any; } {
+    return {
+      ami: config.amiId,
+      amiType:
+        { unixData: { sshPort: '22' } },
+      associatePublicIp: false,
+      connectBySSHProcess: false,
+      connectionStrategy: 'PRIVATE_IP',
+      customDeviceMapping: '/dev/xvda=:100:true:::encrypted',
+      deleteRootOnTermination: true,
+      description: `jenkinsAgentNode-${config.workerLabelString}`,
+      ebsEncryptRootVolume: 'ENCRYPTED',
+      ebsOptimized: false,
+      hostKeyVerificationStrategy: 'OFF',
+      iamInstanceProfile: this.AgentNodeInstanceProfileArn,
+      idleTerminationMinutes: '30',
+      initScript: config.initScript,
+      labelString: config.workerLabelString,
+      launchTimeoutStr: '300',
+      maxTotalUses: -1,
+      minimumNumberOfInstances: 0,
+      minimumNumberOfSpareInstances: 1,
+      mode: 'EXCLUSIVE',
```

Comment:
What is this exclusive means here?

Reply:
Looks like node will only be used when that node label is specified: https://github.com/jenkinsci/ec2-plugin/blob/ec2-1.38/src/main/java/hudson/plugins/ec2/SlaveTemplate.java#L1172-L1176
Not sure but most probably it maps to this setting:
<img width="875" alt="image" src="https://user-images.githubusercontent.com/61760125/159982636-3bd04b18-d414-4949-aea4-7a8f30265548.png">

Reply:
Oh you mean this, this is good and should be this way.